devito/operator.py

```diff
@@ -188,6 +196,27 @@ def __init__(self, expressions, **kwargs):
 
     # Compilation
 
+    def _add_implicit(self, expressions):
+        """Create and add any associated implicit expressions."""
+        processed = []
+        for e in expressions:
+            if e.subdomain:
+                try:
+                    dat = e._subdomain._implicit_eq_dat
+                    dims = [d.root for d in e.free_symbols if isinstance(d, Dimension)]
+                    sub_dims = [d.root for d in e.subdomain.dimensions]
+                    dims = list(set(dims).symmetric_difference(set(sub_dims)))
+                    dims.append(e.subdomain._implicit_dimension)
+                    implicit_expressions = [Eq(i['rhs'], i['lhs'], implicit_dims=dims)
+                                            for i in dat]
+                    processed.extend(implicit_expressions)
+                    dims.extend(e.subdomain.dimensions)
+                    e = Eq(e.lhs, e.rhs, subdomain=e.subdomain, implicit_dims=dims)
+                except AttributeError:
+                    pass
+            processed.append(e)
+        return list(dict.fromkeys(processed))
```

Comment:
We need to remove repeat instances of the same implicit equations.

Reply:
so this might break the case in which we have two SubDomainSets which create the same implicit_equation?

if so, to fix it, we have to make sure that the implicit_dimension created within SubDomainSet isn't cached as all Dimensions (as opposed to Functions) do. For example, look [here](https://github.com/opesci/devito/blob/master/tests/test_symbol_caching.py#L208): ``d0`` is actually ``d1``. And SubDomainSet creates a ``Dimension(name='n')``, so it will be eventually shared by multiple SubDomainSets. We have multiple options here again, but the simplest one is to enforce uniqueness by name

```
class SubDomainSet(object):

    _counter = 0
    """docstring here."""

    ...
    Dimension(name='n%d' % SubDomainSet._counter)
    SubDomainSet._counter += 1
```

I'm not a big fan of this approach because now if you execute tests (or Operators, in general) in different order you get different generated code (eg, n0 in one Operator and n1 in the other, and then swapped), but it's definitely the simplest one. 

Thoughts?